### PR TITLE
fix: example-value-or-externalValue too broad JsonPath expressions

### LINKED ISSUE #883: `example-value-or-externalValue` should only target OAS3
#882 puts under the light a minor issue: `value` or `externalValue` only make sense in the context of OAS3, but the rule `example-value-or-externalValue` apply to both OAS2 and OAS3.

#882 could be merged without incurring any breaking change. However, maybe should we rename the rule to `examples-value-or-externalValue` (and update the doco accordingly) to make it clear what it actually targets.

### PR DESCRIPTION
**Pull request by m-mohr**
Related to #883 (partially, see comments).

**Checklist**

- [x] Tests added / updated
- [ ] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

**Additional context**

$..examples is too broad and incorrectly complains about examples that contain a property examples, see https://github.com/stoplightio/spectral/issues/883#issuecomment-571580864

Please some one double-check whether I found all paths for the examples in the OAS3 spec. Not sure about Swagger2?!

Edit: I'm not sure why the CI fails. Is it because some JSON path expressions return an empty array? If yes, the old JSON path would also fail on OpenAPI specs with no examples, right?

### CODE REVIEW

**Comment by nulltoken** [comment]
@m-mohr Could you please add a test that would only pass with that change ("examples that contain a property examples") for instance?

File: src/rulesets/oas/index.json [hunk 1]
```diff
@@ -115,7 +115,12 @@
       "description": "Example should have either a `value` or `externalValue` field.",
       "recommended": true,
       "type": "style",
-      "given": "$..examples.*",
+      "given": [
```
Comment (nulltoken) [inline src/rulesets/oas/index.json]: @m-mohr This rule currently applies to both OAS2 and OAS3.

Could you please double check that all the new paths make sense in both context (cf. https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v2.0/schema.json and https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v3.0/schema.json).

Wouldn't that be the case, an approach would be to create two new rules (prefixed `oas2-` and `oas3-`) that would only target the relevant paths considering the spec version.

Comment (nulltoken) [reply src/rulesets/oas/index.json]: @P0lip would dropping a rule and creating two new ones be considered as a breaking change?

Comment (P0lip) [reply src/rulesets/oas/index.json]: I'd rather avoid creating a new rule if we can help it, but if we *really* need to introduce a separate rule, that wouldn't necessarily be a breaking change.
Is there any way we can leverage multiple paths to target relevant parts of the document?

Comment (m-mohr) [reply src/rulesets/oas/index.json]: @nulltoken Are you sure this applies to OAS2? I can't find any info on examples with value or externalValue in the OAS2 spec. 🤔

Comment (nulltoken) [reply src/rulesets/oas/index.json]: @m-mohr Sorry, I was unclear. Let me try and reword my thoughts: The currently merged rule applies to both OAS2 and OAS3 (with the `"$..examples.*"` JsonPath expression).

I was requesting you to double check that all the new paths this PR adds also apply to OAS2 and OAS3. If that's not the case, then I fear we'll have to eventually have to end up with two rules: one for OAS2 and one for OAS3, each of them leveraging the `format` rule attribute as a target specifier.

Comment (nulltoken) [reply src/rulesets/oas/index.json]: @m-mohr Duh. My brain just exploded for a minute. 😜 

You're right. 

Although the path applies to both specs, the rule only makes sense in OAS3 context (as hinted in #883).

How about about renaming that rule to `oas3-example-value-or-externalValue` and adding a `formats` specifier to make sure it only considers OAS3 documents?

Comment (m-mohr) [reply src/rulesets/oas/index.json]: Sure, would make sense, but is a breaking change. So I'd probably prefer to merge this as non-breaking change to get the fix faster into a release and then change the name for the next major release.

**Comment by m-mohr** [comment]
@nulltoken I could post an example here, but I doubt I can add a good test. I'm not really into your test suite and don't realky know what to do there. Could someone make a test out if a json example I post? Or would a new scenario file be enough? 🤔

File: src/rulesets/oas/index.json (hunk 1 above)
Comment (nulltoken) [reply src/rulesets/oas/index.json]: >So I'd probably prefer to merge this as non-breaking change to get the fix faster into a release

@m-mohr :+1: Let's deal with this later as part of #883.

**Comment by nulltoken** [comment]
>Or would a new scenario file be enough?

@m-mohr That would be perfect!

**Comment by nulltoken** [comment]
@m-mohr The tests are currently failing because the test documents (incorrectly) expose the `examples` at the root of the document. They were previously passing because of the (too) broad JsonPath expression.

Could you please update them so that they pass? This would require simple changes in the test documents and expectations.

>I'm not really into your test suite and don't really know what to do there. 

An example below:

```diff
diff --git a/src/rulesets/oas/__tests__/example-value-or-externalValue.ts b/src/rulesets/oas/__tests__/example-value-or-externalValue.ts
index 796ccc0..820f547 100644
--- a/src/rulesets/oas/__tests__/example-value-or-externalValue.ts
+++ b/src/rulesets/oas/__tests__/example-value-or-externalValue.ts
@@ -56,21 +56,21 @@ describe('example-value-or-externalValue', () => {
 
   test('multiple examples - return warnings if missing externalValue and value in one', async () => {
     const results = await s.run({
-      examples: { first: { value: 'value1' }, second: { externalValue: 'external-value2' }, third: {} },
+      components: { examples: { first: { value: 'value1' }, second: { externalValue: 'external-value2' }, third: {} } },
     });
     expect(results).toEqual([
       {
         code: 'example-value-or-externalValue',
         message: 'Example should have either a `value` or `externalValue` field.',
-        path: ['examples', 'third'],
+        path: ['components', 'examples', 'third'],
         range: {
           end: {
-            character: 15,
-            line: 8,
+            character: 17,
+            line: 9,
           },
           start: {
-            character: 12,
-            line: 8,
+            character: 14,
+            line: 9,
           },
         },
         severity: DiagnosticSeverity.Warning,
```

**Note:** in order to repro those failure locally, running `yarn test "example-value"` should work.

**Comment by m-mohr** [comment]
@nulltoken Thanks for the hint. I updated the tests as requested and added new tests, also checking the
* other paths (parameters, headers, content)
* examples on top level
* examples properties in examples

So I think I made all your requested changes.

**Comment by nulltoken** [comment]
@m-mohr Looks very good to me.

@P0lip How do you feel about this?

**Comment by m-mohr** [comment]
Great!
Why's the `stoplightio.spectral` check hanging?

**Comment by P0lip** [comment]
> Why's the stoplightio.spectral check hanging?

Most likely some permissions issue, as you don't belong to Stoplight organization.
@XVincentX any clue? It's Azure Pipelines.

File: src/rulesets/oas/index.json [hunk 2]
```diff
@@ -115,7 +115,12 @@
...
+        "$..content.*.examples.*",
```
Comment (P0lip) [inline src/rulesets/oas/index.json]: Any chance we could narrow these down?
`$..` path expressions are fairly expensive, as we need to traverse deeply pretty much the whole object.
If we could have something more or less close to
`$.paths.*..parameters` and `$.components..parameters` that would be already a huge step forward.

Same for `content` and `headers`. If it happens to be very troublesome to implement, I can live with it.

Comment (m-mohr) [reply src/rulesets/oas/index.json]: It's just a matter of going through the spec and compiling all cases. I'd expect that it blows up the list quite a lot . Also, I'd not be sure whether I found all cases and I would expect anybody to carefully double-check it. So I'll only do the work if anybody would agree on checking it.

Comment (m-mohr) [reply src/rulesets/oas/index.json]: I just tried to resolve `"$..content.*.examples.*",` and it seems to become 
```
        "$.paths.*.*.requestBody.content.*.examples.*",
	"$.paths.*.*.callbacks.*.*.requestBody.content.*.examples.*",
        "$.components.callbacks.*.*.*.requestBody.content.*.examples.*",
        "$.components.requestBodies.*.content.*.examples.*",
```
BUT: I'm not at all sure whether that's correct, but I tried my best. And I think that was the easiest case. Also, some of the * could be replaced by (get or post or patch or put or ...), but I'm not sure how OR works in JSON Path without blowing the expression to be overly long and complicated. So I think I'd not be able to provide a proper list at the moment, time is too limited.

Comment (m-mohr) [reply src/rulesets/oas/index.json]: I committed the version above, the tests succeed.

Comment (P0lip) [reply src/rulesets/oas/index.json]: @m-mohr yeah, that's much better. Thanks!

**Comment by XVincentX** [comment]
Checking this out @P0lip

**Comment by XVincentX** [comment]
@P0lip For some reason the build was not triggered — some web hook failed. I've run this manually and it passed the tests, we can just merge it.

**Comment by XVincentX** [comment]
Ah ok — I think the reason is that this is an external PR.

**Pull request merged**